### PR TITLE
fix(structured-list): add horizontal overflow scrolling

### PR DESCRIPTION
**Pull request by andy-blum**
### Related Ticket(s)

Relates to [ADCMS-2768](https://jsw.ibm.com/browse/ADCMS-2768):

> The Structured list component used for the sections "Compare editions" is shown ok on the desktop version but is not mobile friendly at all, only the first column can be seen and there is no visible or functional way to see the content "hidden" on the right side of the page. 
> 
> URL: https://www.ibm.com/products/flashsystem-5000#container-932d270675

### Description

This PR adds a horizontal scroll functionality to the structured-list and pricing-table components. The scroll bar is hidden with CSS and a combination of `scroll-snap` properties and `Element.scrollBy()` methods allow us to smoothly transition back and forth between table sections.

### Changelog

**New**

- Add overflow scrolling to structured-list and pricing-table components

### CODE REVIEW

**Review by jkaeser — CHANGES_REQUESTED** [review]
This is looking really great for Structured List! I didn't know about the `scroll-snap-` stuff you've made use of, and I like the end result.

The biggest issue I can see is that this breaks the Pricing Table's sticky header. It relies on `position: sticky;`, which appears to only work when there is no overflow on its parent.

File: packages/styles/scss/components/structured-list/_structured-list.scss [hunk 1]
```diff
@@ -167,14 +168,68 @@
...
+
+      &.right {
+        left: calc(100%);
+        transform: translateX(1rem);
+
+        &::before {
+          right: 0;
+          background-image: linear-gradient(90deg, transparent, $ui-01);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (andy-blum) [reply packages/styles/scss/components/structured-list/_structured-list.scss:220]: I started attempting to make this work for LTR and RTL languages, but the mechanism in place here _does break in RTL_.

However, with Arabic relegated to the "unsupported" language category, I'm not certain that matters. Would love input from @kennylam on if we need to support RTL in all components going forward.

[the remark this answers is not in the document]
Comment (andy-blum) [reply packages/styles/scss/components/structured-list/_structured-list.scss:220]: Demo of this mechanism not working in RTL: https://codepen.io/andy-blum/pen/WNyYvRQ

File: packages/web-components/src/components/structured-list/structured-list.ts [hunk 2]
```diff
@@ -67,6 +104,15 @@ class DDSStructuredList extends StableSelectorMixin(LitElement) {
   }
 
   render() {
+    const { wrapperId } = this.constructor as typeof DDSStructuredList;
```
Comment (jkaeser) [inline packages/web-components/src/components/structured-list/structured-list.ts:107]: Looks like you need to get rid of this unused variable to fix a CI error.

**Review by jkaeser — CHANGES_REQUESTED** [review]
@andy-blum I'm not seeing any new changes that might fix the Pricing Table. Do you think we should take care of that in a separate issue?

**Comment by andy-blum** [comment]
@jkaeser yeah, I don't think we'll be able to keep the sticky header _and_ make the table responsive. We probably need to bring this up in a design/dev meeting after the first.

**Comment by jwitkowski79** [comment]
@andy-blum Hello! I have a couple of questions:
1. The web component link in the PR only shows the structured list. Was this fix also applied to pricing tables? If so, is there a link where I could also review the change to pricing tables?
2. On my mobile device, I'm noticing that the horizontal scroll bar is hidden until the user interacts with the table. I'm curious if you think that this would be an accessibility issue? Depending on the content, the gradient to the right might not be as noticeable and I'm thinking that maybe we don't hide the scroll bar to provide an extra additional prompt that the table does scroll horizontally. Curious to get your thoughts on this.

**Comment by dratzenboeck-ibm** [comment]
FYI there was an issue reported where the sticky header of the TOC is not working on mobile when the structured list is on the page: https://www.ibm.com/flashsystem . Would this also be part of this PR?

**Comment by andy-blum** [comment]
@jwitkowski79 

> The web component link in the PR only shows the structured list. Was this fix also applied to pricing tables? If so, is there a link where I could also review the change to pricing tables?

Yes, the fix should be present for pricing tables, I'm attaching some gifs of both components in action, and have updated the storybook styles to hopefully make the change a bit more evident. The problem that remains is the pricing table's sticky headers are no longer sticky headers, and will scroll away like normal elements.

structured-list
pricing-table

> On my mobile device, I'm noticing that the horizontal scroll bar is hidden until the user interacts with the table. I'm curious if you think that this would be an accessibility issue? Depending on the content, the gradient to the right might not be as noticeable and I'm thinking that maybe we don't hide the scroll bar to provide an extra additional prompt that the table does scroll horizontally. Curious to get your thoughts on this.

The scrollbar comes from the OS/browser combo, and we don't have any real control over that. We could maybe make the gradient more obvious, but then we have different overflow gradients across the design system. I'll leave that decision to you.

@dratzenboeck-ibm 

> an issue reported where the sticky header of the TOC is not working on mobile when the structured list is on the page

I highly doubt these two issues are related, but I'll investigate a bit more.

**Comment by proeung** [comment]
> The problem that remains is the pricing table's sticky headers are no longer sticky headers, and will scroll away like normal elements.

I just wanted to follow up on this PR to see if we can come up with a solution to the sticky behavior or at least a middle-ground solution moving forward that satisfies the Interfaces team (@jwitkowski79 @jwdomb). @jkaeser I'd like to get your input on this since you've implemented the sticky header behavior for the pricing table.

**Issue:** Currently, the horizontal scroll function is not working on Pricing Table and Structure List components because the production/AEM page has an `overflow-x: hidden;` property that's blocking the horizontal scroll from working (see attached).


- **Solution Option 1**: Take the solution in this PR (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/9713/files#diff-4e80dc6063d7cc3b405f2301c8664f49deaf6b752ccbb5b16f5f4154eebe81f1R177-R179) that makes the Pricing Table and Structure List component scrollable, however, disable the sticky header behavior for the Pricing Table on mobile. This is a good short-term solution for the scroll issue that we're seeing on production.
- **Solution Option 2 (More engineering effort)**: Make the Pricing Table component scrollable and adjust when the table header should be collapsible and get an updated offset calculation (see attached).

-------

https://user-images.githubusercontent.com/1815714/220166576-90a1c987-56d2-4f06-8164-71ecb52aae90.mov

-------


AEM test page with Pricing Table header (_might be reset after release deployment_): https://stage-author.np-roks.cms.cis.ibm.net/content/adobe-cms/us/en/regression-testing-1-28-0/new-template.html?wcmmode=disabled

**Comment by jkaeser** [comment]
> * **Solution Option 2 (More engineering effort)**: Make the Pricing Table component scrollable and adjust when the table header should be collapsible and get an updated offset calculation (see attached).

@proeung I'm not sure what I'm looking at in the attached video. Is that a bug currently in AEM?

I think the long-term solution has to move away from relying on `position: sticky;` for the Pricing Table header's stickiness. Despite giving us the exact behavior we want with relatively little manual calculation, the fact that this style rule can be broken by any DOM ancestor element using an incompatible `overflow` rule makes the component super brittle.

Is this is alignment with what you're thinking?

**Comment by jkaeser** [comment]
> * **Solution Option 1**: Take the solution in this PR (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/9713/files#diff-4e80dc6063d7cc3b405f2301c8664f49deaf6b752ccbb5b16f5f4154eebe81f1R177-R179) that makes the Pricing Table and Structure List component scrollable, however, disable the sticky header behavior for the Pricing Table on mobile. This is a good short-term solution for the scroll issue that we're seeing on production.

@andy-blum Is it possible to only apply the `overflow` styles on mobile once these tables become eligible for horizontal scrolling? That should allow us to preserve the Pricing Table's sticky header until mobile, right?

**Comment by andy-blum** [comment]
> Is it possible to only apply the overflow styles on mobile once these tables become eligible for horizontal scrolling? That should allow us to preserve the Pricing Table's sticky header until mobile, right?

🤔 that's an interesting thought. I wonder what @jwitkowski79 would think about this.

**Comment by proeung** [comment]
2/27 call with @andy-blum @jwitkowski79 @proeung summary:

- Move forward with Solution **Option 1** (see - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/9713#issuecomment-1437340943)
  - Implement a fix for the overflow scroll issue seen on mobile and disable the sticky behavior within the pricing table on all breakpoints, until we have a better long-term solution for the sticky navigation.
  - If a Pricing Table showcases CTAs and has a long table, we're going to recommend that the Content Authors repeat the CTA to the bottom of the table so that the user can still scroll down the list and still be able to click on the CTAs without scroll back up to the top.
  - Follow up ticket will be created to look into moving away from relying on `position: sticky;` for the Pricing Table header's stickiness and test how it can be best integrated into AEM given that it has certain overflow rules that might affect the sticky behavior.
  
**Code deployment and target date:**

- The AEM production team is looking to migrate a bunch of PLG products into AEM in the next few weeks and these pages include pricing tables. Given the current ibm.com release schedule (see - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/9698), the changes coming out of this PR will most likely go into the `v1.45.0` which has a full release date of **Mar 28, 2023**. Additionally, we also have to factor in the AEM deployment schedule, which is another 2 weeks out.
- Keeping the PLG migration in mind, we think that it's probably best to implement a stop-gap solution for the horizontal scroll behavior issue on the AEM side, where we add in the `overflow: auto;` onto the div used to wrap the Structure List and Pricing Table components (see attached).
- This workaround solution can be part of the next AEM deployment (Version AEM 1.30.0 - Isabelle), which is around **03/14/23**
- Follow up ticket will be created to track the effort on the AEM side.
- We'll continue to implement the long-term solution in this upstream repo.


**Video screenshot of the issue on Chrom's Mobile simulator**
https://user-images.githubusercontent.com/1815714/221694010-62dd670f-7101-40cc-9b54-b6f21f8d5e45.mov


**Expected behavior/fix once `overflow:auto;` is applied to the `.structured-list--container` class**
https://user-images.githubusercontent.com/1815714/221694047-2a4bea02-b5f1-4ec0-9728-8ee9bf8d033e.mov

**Review by proeung — APPROVED** [review]
Looks good! Thanks for applying the changes based on what we discussed.

@jwitkowski79 Can you give this PR another review? Pricing Table without fixed header - https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/9713/index.html?path=/story/components-pricing-table--without-row-headers

**Comment by andy-blum** [comment]
After discussion with @jwitkowski79 and trying to fix that rouge scrollbar issue she found, we're going to submit this PR as-is and open a followup if needed.

**Review by jwitkowski79 — APPROVED** [review]
@andy-blum approving this under the stipulation that we're going to open up a separate issue to investigate the rouge scrollbar. Thanks for your work on this!

**Comment by proeung** [comment]
@jwitkowski79 @andy-blum Sounds good! Perhaps, we can highlight the rouge scrollbar in this follow-up ticket (https://jsw.ibm.com/browse/ADCMS-3020) or open a new one, if needed.

@kennylam or @annawen1 Can we get your review/approval for this PR so that the changes can be merged and slated for the `v1.45.0` release cut? Thanks!

**Review by annawen1 — APPROVED** [review]

**Pull request merged**